lib/upgrade/etcd/conversionv1v3/workloadendpoint.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/names"
+	"github.com/projectcalico/libcalico-go/lib/net"
+)
+
+type WorkloadEndpoint struct{}
+
+// APIV1ToBackendV1 converts v1 WorkloadEndpoint API to v1 WorkloadEndpoint KVPair.
+func (_ WorkloadEndpoint) APIV1ToBackendV1(rIn unversioned.Resource) (*model.KVPair, error) {
+	ah := rIn.(apiv1.WorkloadEndpoint)
+	k, err := convertMetadataToKey(ah.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// IP networks are stored in the datastore in separate IPv4 and IPv6
+	// fields.  We normalise the network to ensure the IP is correctly
+	// masked.
+	ipv4Nets := []net.IPNet{}
+	ipv6Nets := []net.IPNet{}
+	for _, n := range ah.Spec.IPNetworks {
+		n = *(n.Network())
+		if n.Version() == 4 {
+			ipv4Nets = append(ipv4Nets, n)
+		} else {
+			ipv6Nets = append(ipv6Nets, n)
+		}
+	}
+
+	ipv4NAT := []model.IPNAT{}
+	ipv6NAT := []model.IPNAT{}
+	for _, n := range ah.Spec.IPNATs {
+		nat := model.IPNAT{IntIP: n.InternalIP, ExtIP: n.ExternalIP}
+		if n.InternalIP.Version() == 4 {
+			ipv4NAT = append(ipv4NAT, nat)
+		} else {
+			ipv6NAT = append(ipv6NAT, nat)
+		}
+	}
+
+	var ports []model.EndpointPort
+	for _, port := range ah.Spec.Ports {
+		ports = append(ports, model.EndpointPort{
+			Name:     port.Name,
+			Protocol: port.Protocol,
+			Port:     port.Port,
+		})
+	}
+
+	d := model.KVPair{
+		Key: k,
+		Value: &model.WorkloadEndpoint{
+			Labels:           ah.Metadata.Labels,
+			ActiveInstanceID: ah.Metadata.ActiveInstanceID,
+			State:            "active",
+			Name:             ah.Spec.InterfaceName,
+			Mac:              ah.Spec.MAC,
+			ProfileIDs:       ah.Spec.Profiles,
+			IPv4Nets:         ipv4Nets,
+			IPv6Nets:         ipv6Nets,
+			IPv4NAT:          ipv4NAT,
+			IPv6NAT:          ipv6NAT,
+			IPv4Gateway:      ah.Spec.IPv4Gateway,
+			IPv6Gateway:      ah.Spec.IPv6Gateway,
+			Ports:            ports,
+		},
+		Revision: ah.Metadata.Revision,
+	}
+
+	log.Debugf("Converted: %+v\n To: %+v", ah, d)
+
+	return &d, nil
+}
+
+// BackendV1ToAPIV3 converts v1 WorkloadEndpoint KVPair to v3 API.
+func (_ WorkloadEndpoint) BackendV1ToAPIV3(kvp *model.KVPair) (Resource, error) {
+	wepKey, ok := kvp.Key.(model.WorkloadEndpointKey)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid WorkloadEndpoint resource key")
+	}
+	wepValue, ok := kvp.Value.(*model.WorkloadEndpoint)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid WorkloadEndpoint resource Value")
+	}
+
+	labels := convertLabels(wepValue.Labels)
+	namespace := "default"
+
+	var pod string
+	var container string
+	var workload string
+
+	// Populate our values based on the orchestrator.
+	switch wepKey.OrchestratorID {
+	case "k8s":
+		namespace, pod = getPodNamespaceName(wepKey.WorkloadID)
+		workload = convertName(wepKey.WorkloadID)
```

Comment:
I think for k8s, we want workload to be an empty string, so just remove this line.